File: MicrostockMetadata/eps_renderer.py
import os
import hashlib
import struct
import subprocess
import tempfile
import base64

_eps_cache = {}


def clean_eps_buffer(raw_bytes):
    if len(raw_bytes) >= 30:
        if raw_bytes[0] == 0xC5 and raw_bytes[1] == 0xD0 and raw_bytes[2] == 0xD3 and raw_bytes[3] == 0xC6:
            ps_offset = struct.unpack_from('<I', raw_bytes, 4)[0]
            ps_length = struct.unpack_from('<I', raw_bytes, 8)[0]
            if ps_offset > 0 and ps_length > 0 and ps_offset + ps_length <= len(raw_bytes):
                return raw_bytes[ps_offset:ps_offset + ps_length]
    return raw_bytes


def parse_eps_bounding_box(raw_bytes):
    try:
        text_header = raw_bytes[:min(len(raw_bytes), 32768)].decode('latin1', errors='replace')
        import re
        bbox_match = (
            re.search(r'%%BoundingBox:\s*(-?\d+)\s+(-?\d+)\s+(\d+)\s+(\d+)', text_header, re.IGNORECASE)
            or re.search(r'%%HiResBoundingBox:\s*([\d\.]+)\s+([\d\.]+)\s+([\d\.]+)\s+([\d\.]+)', text_header, re.IGNORECASE)
        )
        if bbox_match:
            x1, y1 = float(bbox_match.group(1)), float(bbox_match.group(2))
            x2, y2 = float(bbox_match.group(3)), float(bbox_match.group(4))
            w, h = abs(x2 - x1), abs(y2 - y1)
            if w > 10 and h > 10:
                max_dim_pt = max(w, h)
                target_px = 1800
                calculated_dpi = round((target_px / max_dim_pt) * 72)
                clamped_dpi = max(100, min(300, calculated_dpi))
                return {'width': round(w), 'height': round(h), 'dpi': clamped_dpi}
    except Exception:
        pass
    return {'width': 800, 'height': 600, 'dpi': 200}


def validate_png_buffer(buf):
    if not buf or len(buf) < 24:
        return {'valid': False, 'width': 0, 'height': 0, 'error': 'PNG buffer is too short or empty.'}
    is_png = (
        buf[0] == 0x89 and buf[1] == 0x50 and buf[2] == 0x4E and buf[3] == 0x47
        and buf[4] == 0x0D and buf[5] == 0x0A and buf[6] == 0x1A and buf[7] == 0x0A
    )
    if not is_png:
        return {'valid': False, 'width': 0, 'height': 0, 'error': 'Rendered file is not a valid PNG image.'}
    width = struct.unpack_from('>I', buf, 16)[0]
    height = struct.unpack_from('>I', buf, 20)[0]
    if width <= 0 or height <= 0:
        return {'valid': False, 'width': width, 'height': height, 'error': 'Rendered PNG has invalid dimensions.'}
    return {'valid': True, 'width': width, 'height': height}


def render_eps_to_png(base64_data, filename='artwork.eps'):
    clean_b64 = base64_data.split(',')[1] if ',' in base64_data else base64_data
    raw_bytes = base64.b64decode(clean_b64)

    if not raw_bytes or len(raw_bytes) == 0:
        raise ValueError('Uploaded EPS file is empty or corrupted.')

    file_hash = hashlib.sha256(raw_bytes).hexdigest()
    if file_hash in _eps_cache:
        return _eps_cache[file_hash]

    clean_buffer = clean_eps_buffer(raw_bytes)
    bbox = parse_eps_bounding_box(clean_buffer)
    dpi = bbox['dpi']

    tmp_dir = tempfile.gettempdir()
    file_id = f"eps_{os.getpid()}_{hashlib.md5(raw_bytes[:1024]).hexdigest()[:12]}"
    eps_path = os.path.join(tmp_dir, f"{file_id}.eps")
    png_path = os.path.join(tmp_dir, f"{file_id}.png")

    try:
        with open(eps_path, 'wb') as f:
            f.write(clean_buffer)

        render_success = False

        try:
            subprocess.run(
                [
                    'gs', '-dSAFER', '-dBATCH', '-dNOPAUSE', '-dEPSCrop',
                    '-sDEVICE=pngalpha', f'-r{dpi}',
                    '-dTextAlphaBits=4', '-dGraphicsAlphaBits=4',
                    f'-sOutputFile={png_path}', eps_path,
                ],
                check=True, timeout=15,
                capture_output=True,
            )
            if os.path.exists(png_path):
                render_success = True
        except (FileNotFoundError, subprocess.CalledProcessError, subprocess.TimeoutExpired):
            pass

        if not render_success:
            try:
                subprocess.run(
                    [
                        'convert', '-density', str(dpi), '-colorspace', 'sRGB',
                        eps_path, '-resize', '2000x2000>', png_path,
                    ],
                    check=True, timeout=15,
                    capture_output=True,
                )
                if os.path.exists(png_path):
                    render_success = True
            except (FileNotFoundError, subprocess.CalledProcessError, subprocess.TimeoutExpired):
                pass

        if not render_success or not os.path.exists(png_path):
            raise ValueError('Unable to render EPS preview. Please retry.')

        with open(png_path, 'rb') as f:
            png_buffer = f.read()

        validation = validate_png_buffer(png_buffer)
        if not validation['valid']:
            raise ValueError(validation.get('error', 'Unable to render EPS preview.'))

        pixel_width = validation['width']
        pixel_height = validation['height']

        ratio = pixel_width / pixel_height if pixel_height > 0 else 1
        if abs(ratio - 1) < 0.08:
            orientation = 'Square'
        elif ratio < 0.92:
            orientation = 'Portrait'
        else:
            orientation = 'Landscape'

        b64_result = base64.b64encode(png_buffer).decode('utf-8')
        preview_url = f'data:image/png;base64,{b64_result}'

        result = {
            'previewUrl': preview_url,
            'base64Data': b64_result,
            'mimeType': 'image/png',
            'width': pixel_width,
            'height': pixel_height,
            'dimensions': f'{pixel_width} x {pixel_height} px (EPS Vector)',
            'orientation': orientation,
            'hasTransparency': True,
            'backgroundType': 'Transparent',
            'dominantColors': ['#38bdf8', '#818cf8'],
        }

        _eps_cache[file_hash] = result
        return result

    finally:
        for p in (eps_path, png_path):
            try:
                if os.path.exists(p):
                    os.unlink(p)
            except OSError:
                pass

File: microstock-metadata-ai/server/epsRenderer.ts
import fs from 'fs';
import path from 'path';
import crypto from 'crypto';
import { execFile } from 'child_process';
import { promisify } from 'util';

const execFileAsync = promisify(execFile);

// Cache: SHA-256 of EPS content -> Rendered PNG Preview Data
interface CachedEpsPreview {
  previewUrl: string;
  base64Data: string;
  mimeType: string;
  width: number;
  height: number;
  dimensions: string;
  orientation: 'Square' | 'Portrait' | 'Landscape';
  hasTransparency: boolean;
  backgroundType: 'Transparent' | 'White' | 'Black' | 'Colored' | 'Gradient' | 'Complex';
  dominantColors: string[];
}

const epsCache = new Map<string, CachedEpsPreview>();

/**
 * Strips DOS EPS binary header (0xC5D0D3C6) to extract pure PostScript stream
 */
export function cleanEpsBuffer(buffer: Buffer): Buffer {
  if (buffer.length >= 30 && buffer[0] === 0xc5 && buffer[1] === 0xd0 && buffer[2] === 0xd3 && buffer[3] === 0xc6) {
    const psOffset = buffer.readUInt32LE(4);
    const psLength = buffer.readUInt32LE(8);
    if (psOffset > 0 && psLength > 0 && psOffset + psLength <= buffer.length) {
      return buffer.subarray(psOffset, psOffset + psLength);
    }
  }
  return buffer;
}

/**
 * Parse BoundingBox from EPS headers to estimate dimensions & aspect ratio
 */
export function parseEpsBoundingBox(buffer: Buffer): { width: number; height: number; dpi: number } {
  try {
    const textHeader = buffer.toString('latin1', 0, Math.min(buffer.length, 32768));
    const bboxMatch =
      textHeader.match(/%%BoundingBox:\s*(-?\d+)\s+(-?\d+)\s+(\d+)\s+(\d+)/i) ||
      textHeader.match(/%%HiResBoundingBox:\s*([\d\.]+)\s+([\d\.]+)\s+([\d\.]+)\s+([\d\.]+)/i);

    if (bboxMatch) {
      const x1 = parseFloat(bboxMatch[1]);
      const y1 = parseFloat(bboxMatch[2]);
      const x2 = parseFloat(bboxMatch[3]);
      const y2 = parseFloat(bboxMatch[4]);
      const w = Math.abs(x2 - x1);
      const h = Math.abs(y2 - y1);
      if (w > 10 && h > 10) {
        const maxDimPt = Math.max(w, h);
        // Target ~1800px on max dimension (between 1600px - 2400px)
        const targetPx = 1800;
        const calculatedDpi = Math.round((targetPx / maxDimPt) * 72);
        const clampedDpi = Math.max(100, Math.min(300, calculatedDpi));
        return { width: Math.round(w), height: Math.round(h), dpi: clampedDpi };
      }
    }
  } catch (e) {
    console.warn('BoundingBox parse warning:', e);
  }
  return { width: 800, height: 600, dpi: 200 };
}

/**
 * Validate PNG buffer signature and read width/height from IHDR chunk
 */
export function validatePngBuffer(buffer: Buffer): { valid: boolean; width: number; height: number; error?: string } {
  if (!buffer || buffer.length < 24) {
    return { valid: false, width: 0, height: 0, error: 'PNG buffer is too short or empty.' };
  }

  // Check PNG signature 89 50 4E 47 0D 0A 1A 0A
  const isPng =
    buffer[0] === 0x89 &&
    buffer[1] === 0x50 &&
    buffer[2] === 0x4e &&
    buffer[3] === 0x47 &&
    buffer[4] === 0x0d &&
    buffer[5] === 0x0a &&
    buffer[6] === 0x1a &&
    buffer[7] === 0x0a;

  if (!isPng) {
    return { valid: false, width: 0, height: 0, error: 'Rendered file is not a valid PNG image.' };
  }

  // Read IHDR width and height (Big Endian at offset 16 and 20)
  const width = buffer.readUInt32BE(16);
  const height = buffer.readUInt32BE(20);

  if (width <= 0 || height <= 0) {
    return { valid: false, width, height, error: 'Rendered PNG has invalid dimensions (0 width or height).' };
  }

  return { valid: true, width, height };
}

/**
 * Render EPS Buffer to PNG Preview using Ghostscript or ImageMagick
 */
export async function renderEpsToPng(
  rawBuffer: Buffer,
  fileName: string = 'artwork.eps'
): Promise<CachedEpsPreview> {
  // 1. Calculate SHA-256 hash for caching
  const hash = crypto.createHash('sha256').update(rawBuffer).digest('hex');
  const cached = epsCache.get(hash);
  if (cached) {
    return cached;
  }

  const cleanBuffer = cleanEpsBuffer(rawBuffer);
  const { width: ptW, height: ptH, dpi } = parseEpsBoundingBox(cleanBuffer);

  const tmpDir = '/tmp';
  const fileId = `eps_${Date.now()}_${crypto.randomBytes(6).toString('hex')}`;
  const epsPath = path.join(tmpDir, `${fileId}.eps`);
  const pngPath = path.join(tmpDir, `${fileId}.png`);

  try {
    await fs.promises.writeFile(epsPath, cleanBuffer);

    let renderSuccess = false;
    let lastError: any = null;

    // Attempt 1: Ghostscript
    try {
      const gsArgs = [
        '-dSAFER',
        '-dBATCH',
        '-dNOPAUSE',
        '-dEPSCrop',
        '-sDEVICE=pngalpha',
        `-r${dpi}`,
        '-dTextAlphaBits=4',
        '-dGraphicsAlphaBits=4',
        `-sOutputFile=${pngPath}`,
        epsPath,
      ];

      await execFileAsync('gs', gsArgs, { timeout: 15000 });
      if (fs.existsSync(pngPath)) {
        renderSuccess = true;
      }
    } catch (gsErr) {
      lastError = gsErr;
      console.warn(`[EPS Renderer] Ghostscript failed for ${fileName}, trying ImageMagick fallback...`, gsErr);
    }

    // Attempt 2: ImageMagick Convert Fallback
    if (!renderSuccess) {
      try {
        const convertArgs = [
          '-density',
          `${dpi}`,
          '-colorspace',
          'sRGB',
          epsPath,
          '-resize',
          '2000x2000>',
          pngPath,
        ];
        await execFileAsync('convert', convertArgs, { timeout: 15000 });
        if (fs.existsSync(pngPath)) {
          renderSuccess = true;
        }
      } catch (convErr) {
        lastError = convErr;
        console.error(`[EPS Renderer] ImageMagick also failed for ${fileName}:`, convErr);
      }
    }

    if (!renderSuccess || !fs.existsSync(pngPath)) {
      throw new Error(`Unable to render EPS preview. Please retry.`);
    }

    const pngBuffer = await fs.promises.readFile(pngPath);
    const validation = validatePngBuffer(pngBuffer);

    if (!validation.valid) {
      throw new Error(validation.error || 'Unable to render EPS preview. Please retry.');
    }

    const { width: pixelWidth, height: pixelHeight } = validation;

    let orientation: 'Square' | 'Portrait' | 'Landscape' = 'Landscape';
    const ratio = pixelWidth / pixelHeight;
    if (Math.abs(ratio - 1) < 0.08) {
      orientation = 'Square';
    } else if (ratio < 0.92) {
      orientation = 'Portrait';
    } else {
      orientation = 'Landscape';
    }

    const base64Data = pngBuffer.toString('base64');
    const previewUrl = `data:image/png;base64,${base64Data}`;

    const result: CachedEpsPreview = {
      previewUrl,
      base64Data,
      mimeType: 'image/png',
      width: pixelWidth,
      height: pixelHeight,
      dimensions: `${pixelWidth} × ${pixelHeight} px (EPS Vector)`,
      orientation,
      hasTransparency: true,
      backgroundType: 'Transparent',
      dominantColors: ['#38bdf8', '#818cf8'],
    };

    // Store in cache
    epsCache.set(hash, result);

    return result;
  } finally {
    // Cleanup temporary files
    try {
      if (fs.existsSync(epsPath)) await fs.promises.unlink(epsPath);
    } catch {}
    try {
      if (fs.existsSync(pngPath)) await fs.promises.unlink(pngPath);
    } catch {}
  }
}
